### ISSUE TITLE
4.1, Archlinux templates : xorg-server (21.1.1-2) breaks dependency required by qubes-vm-gui

### ISSUE THREAD

**Issue by jobforacowboy** [QubesOS/qubes-issues#7048]
```
[user@archlinux ~]$ sudo pacman -Suuy
:: Synchronizing package databases...
 qubes-r4.1 is up to date
 core                  137.8 KiB  48.1 KiB/s 00:03 [----------------------] 100%
 extra                1567.7 KiB   911 KiB/s 00:02 [----------------------] 100%
 community               5.8 MiB  2.77 MiB/s 00:02 [----------------------] 100%
 multilib is up to date
:: Starting full system upgrade...
resolving dependencies...
looking for conflicting packages...
error: failed to prepare transaction (could not satisfy dependencies)
:: installing xorg-server (21.1.1-2) breaks dependency 'X-ABI-VIDEODRV_VERSION=24.0' required by qubes-vm-gui
:: installing xorg-server (21.1.1-2) breaks dependency 'X-ABI-XINPUT_VERSION=24.1' required by qubes-vm-gui
[user@archlinux ~]$ 

```

**Comment by james-callahan** [comment]
This is also an issue on 4.0

**Comment by james-callahan** [comment]
FYI I upgraded xorg to 21.1.1 and recompiled the agent (or at least I thought I did. I now have the suspicion that the PKGBUILD just downloads a precompiled binary); and then I was unable to execute anything in my archlinux VMs.
  - opening e.g a terminal wouldn't work
  - `qvm-run` would just hang

**Comment by DemiMarie** [comment]
> FYI I upgraded xorg to 21.1.1 and recompiled the agent (or at least I thought I did. I now have the suspicion that some part of the PKGBUILD just downloads a precompiled binary); and then I was unable to execute anything in my archlinux VMs.
> 
> * opening e.g a terminal wouldn't work
> * `qvm-run` would just hang

What PKGBUILD are you using?

**Comment by james-callahan** [comment]
> What PKGBUILD are you using?

I copied the change from https://github.com/QubesOS/qubes-gui-agent-linux/pull/146/files into https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=qubes-gui-agent-linux

**Comment by lubellier** [comment]
On 4.1, with qubes-builder I built the `qubes-vm-gui` package with the @ejose19 's [PR146](https://github.com/QubesOS/qubes-gui-agent-linux/pull/146).  The installed `qubes-vm-gui` package fixed this issue. Thanks @ejose19.

@james-callahan : the Qubes-OS team maintains only the PKGBUILD from [qubes-gui-agent-linux](https://github.com/QubesOS/qubes-gui-agent-linux/blob/master/archlinux/PKGBUILD), the package build process uses [qubes builder](https://github.com/Qubes-Community/Contents/blob/master/docs/building/building-archlinux-template.md). For AUR, please report on the [AUR package page](https://aur.archlinux.org/packages/qubes-vm-gui/) form. Thanks.

**Comment by james-callahan** [comment]
> @james-callahan : the Qubes-OS team maintains only the PKGBUILD from [qubes-gui-agent-linux](https://github.com/QubesOS/qubes-gui-agent-linux/blob/master/archlinux/PKGBUILD), the package build process uses [qubes builder](https://github.com/Qubes-Community/Contents/blob/master/docs/building/building-archlinux-template.md). For AUR, please report on the [AUR package page](https://aur.archlinux.org/packages/qubes-vm-gui/) form. Thanks.

You'll see the PKGBUILDs are practically the same, but the AUR one is cleaned up a little to follow best practices: `diff -wu <(curl https://raw.githubusercontent.com/QubesOS/qubes-gui-agent-linux/master/archlinux/PKGBUILD) <(curl https://aur.archlinux.org/cgit/aur.git/plain/PKGBUILD?h=qubes-gui-agent-linux)`
If the issue is present in the AUR version I see no reason for it to be absent in the qubes-builder version.

**Comment by 0xdd7fq** [comment]
howto fix:

```
gpg2 --keyserver hkps://pgp.surf.nl --recv-keys 0x063938BA42CFA724
yay -G qubes-gui-agent-linux
cd qubes-gui-agent-linux/
vim PKGBUILD
makepkg -si
```

**Comment by lubellier** [comment]
IMHO, the @ejose19 's https://github.com/QubesOS/qubes-gui-agent-linux/pull/146 PR can now be merged. 
Other related issues :
- https://github.com/QubesOS/qubes-issues/issues/7098
- https://github.com/Qubes-Community/Contents/issues/153

**Comment by seberm** [comment]
Hello,
I am the maintainer of all current `qubes-*` [packages in the AUR](https://aur.archlinux.org/packages/?O=0&K=qubes). I have created them mostly for my personal use, because I think it is easier to install them directly from AUR instead of rebuilding the whole arch template.

I just wanted to clarify that these packages follow the changes in the Qubes. I try to keep the difference at minimal, most of the changes are clean-ups. The updates are done irregularly (mostly when something breaks).

I also tried to test the @jose19's PR (https://github.com/QubesOS/qubes-gui-agent-linux/pull/146). Unfortunately it seems that the change is not working correctly :/ - at least on my Qubes 4.0. I would like to upgrade all `qubes-*` packages to their `v4.1.*` versions after the stable Qubes v4.1 is released, not before.

Does anybody have a workaround which is working for Qubes 4.0?

Thanks,

**Comment by lubellier** [comment]
Hi @seberm ,
  thank for the job done in AUR.
  I'm using Qubes-OS 4.1, and I build myself the packages with the qubes-builder, I applied the ejose19's PR and it worked.
  The ejose19's commit is simple, just the ABI version dependency for Xorg, the same change is simple to backport to [PKGBUILD#L57](https://github.com/QubesOS/qubes-gui-agent-linux/blob/release4.0-staging/archlinux/PKGBUILD#L57) for 4.0.
  See also the @0xdd7fq above comment and the @unman [arch repo](https://qubes.3isec.org/).
  If your problem is about `qubes-gui-agent-linux` and the new Xorg version dependency, **give details** about the error(s). Else, please create a new issue. In each cases, I'll try to help.

**Comment by ejose19** [comment]
@seberm can you please tell me what's the issue with the PR? I made sure to build & install the packages in a vm and they all worked correctly for R4.0. If you mean `git apply` errors, it's due the patch being made against `master` branch following Qubes guidelines (after merge, these are usually backported to R4.0), so you can just simply change the corresponding lines to build for R4.0 in the meanwhile

**Comment by 0xdd7fq** [comment]
@neowutran @unman @icequbes1  can you help to make archlinux template usable on Qubes 4.0/4.1 ?

**Comment by seberm** [comment]
Hello @lubellier , @ejose19 ,
I have done some debugging, the patch is fine.

When I install patched qubes-vm-gui package into (updated) arch template on Qubes v4.0, the VM starts, but it's not able to run any application. The Xorg server is not running at all. It seems the Qubes Xorg modules provided by this package are still compiled for ABI v24 (take a look at logs below).

Does the environment where I create the `qubes-vm-gui` package matter? **The package is created in my current arch template where I still have installed an old xorg-server (v1.20.13-2)**. Maybe that's why the modules are still built for ABI v24.

Content of `~/.xsession-errors`:
```
...
X.Org X Server 1.21.1.1
X Protocol Version 11, Revision 0
Current Operating System: Linux archlinux-test 5.4.156-1.fc25.qubes.x86_64 #1 SMP Fri Oct 29 02:51:34 CEST 2021 x86_64
Kernel command line: root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH rd.plymouth.enable=0 plymouth.enable=0 xen_scrub_pages=0 nopat

Current version of pixman: 0.40.0
...
(==) Log file: "/var/log/Xorg.0.log", Time: Sat Dec 11 19:04:49 2021
(++) Using config file: "/etc/X11/xorg-qubes.conf"
(==) Using system config directory "/usr/share/X11/xorg.conf.d"
(EE) dummyqbs: module ABI major version (24) doesn't match the server's version (25)
(EE) 
Fatal server error:
(EE) no screens found(EE) 
(EE) 
Please consult the The X.Org Foundation support 
	 at http://wiki.x.org
 for help. 
(EE) Please also check the log file at "/var/log/Xorg.0.log" for additional information.
(EE) 
(EE) Server terminated with error (1). Closing log file.
/usr/bin/xinit: giving up
/usr/bin/xinit: unable to connect to X server: Connection refused
/usr/bin/xinit: server error
```

Content of `/var/log/Xorg.0.log`:
```
...
X.Org X Server 1.21.1.1
X Protocol Version 11, Revision 0
[    44.477] Current Operating System: Linux archlinux-test 5.4.156-1.fc25.qubes.x86_64 #1 SMP Fri Oct 29 02:51:34 CEST 2021 x86_64
[    44.477] Kernel command line: root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH rd.plymouth.enable=0 plymouth.enable=0 xen_scrub_pages=0 nopat
[    44.477]  
[    44.477] Current version of pixman: 0.40.0
...
[    44.477] (==) Log file: "/var/log/Xorg.0.log", Time: Sat Dec 11 22:43:25 2021
[    44.477] (++) Using config file: "/etc/X11/xorg-qubes.conf"
[    44.477] (==) Using system config directory "/usr/share/X11/xorg.conf.d"
[    44.477] (==) ServerLayout "Default Layout"
[    44.477] (**) |-->Screen "Screen0" (0)
[    44.477] (**) |   |-->Monitor "Monitor0"
[    44.477] (**) |   |-->Device "Videocard0"
[    44.477] (**) |-->Input Device "qubesdev"
[    44.477] (==) Automatically adding devices
[    44.477] (==) Automatically enabling devices
[    44.477] (==) Automatically adding GPU devices
[    44.477] (==) Automatically binding GPU devices
[    44.477] (==) Max clients allowed: 256, resource mask: 0x1fffff
[    44.477] (WW) The directory "/usr/share/fonts/OTF" does not exist.
[    44.477] 	Entry deleted from font path.
[    44.477] (WW) The directory "/usr/share/fonts/Type1" does not exist.
[    44.477] 	Entry deleted from font path.
[    44.477] (WW) The directory "/usr/share/fonts/100dpi" does not exist.
[    44.477] 	Entry deleted from font path.
[    44.477] (WW) The directory "/usr/share/fonts/75dpi" does not exist.
[    44.478] 	Entry deleted from font path.
[    44.478] (==) FontPath set to:
	/usr/share/fonts/misc,
	/usr/share/fonts/TTF
[    44.478] (==) ModulePath set to "/usr/lib/xorg/modules"
[    44.478] (II) The server relies on udev to provide the list of input devices.
	If no devices become available, reconfigure udev or disable AutoAddDevices.
[    44.478] (II) Module ABI versions:
[    44.478] 	X.Org ANSI C Emulation: 0.4
[    44.478] 	X.Org Video Driver: 25.2
[    44.478] 	X.Org XInput driver : 24.4
[    44.478] 	X.Org Server Extension : 10.0
[    44.478] (++) using VT number 7

[    44.478] (II) systemd-logind: logind integration requires -keeptty and -keeptty was not provided, disabling logind integration
[    44.478] (II) no primary bus or device found
[    44.478] (WW) Open ACPI failed (/var/run/acpid.socket) (No such file or directory)
[    44.478] (II) "glx" will be loaded by default.
[    44.478] (II) LoadModule: "fb"
[    44.478] (II) Module "fb" already built-in
[    44.478] (II) LoadModule: "glx"
[    44.478] (II) Loading /usr/lib/xorg/modules/extensions/libglx.so
[    44.479] (II) Module glx: vendor="X.Org Foundation"
[    44.479] 	compiled for 1.21.1.1, module version = 1.0.0
[    44.479] 	ABI class: X.Org Server Extension, version 10.0

========== !!!
[    44.479] (II) LoadModule: "dummyqbs"
[    44.479] (II) Loading /usr/lib/xorg/modules/drivers/dummyqbs_drv.so
[    44.479] (II) Module dummyqbs: vendor="X.Org Foundation"
[    44.479] 	compiled for 1.20.13, module version = 0.3.6
[    44.479] 	Module class: X.Org Video Driver
[    44.480] 	ABI class: X.Org Video Driver, version 24.1
[    44.480] (EE) dummyqbs: module ABI major version (24) doesn't match the server's version (25)
[    44.480] (EE) Failed to load module "dummyqbs" (module requirement mismatch, 0)
[    44.480] (II) LoadModule: "qubes"
[    44.480] (II) Loading /usr/lib/xorg/modules/drivers/qubes_drv.so
[    44.480] (II) Module qubes: vendor="X.Org Foundation"
[    44.480] 	compiled for 1.20.13, module version = 0.0.1
[    44.480] 	Module class: X.Org XInput Driver
[    44.480] 	ABI class: X.Org XInput driver, version 24.1
[    44.480] (EE) No drivers available.
[    44.480] (EE) 
Fatal server error:
[    44.480] (EE) no screens found(EE) 
========== !!!
...

[    44.480] (EE) Server terminated with error (1). Closing log file.
```

Thanks!

**Comment by lubellier** [comment]
Hi @seberm ,
> Does the environment where I create the qubes-vm-gui package matter? The package is created in my current arch template where I still have installed an old xorg-server (v1.20.13-2). Maybe that's why the modules are still built for ABI v24.

Yes this is important.

Qubes-builder creates a chroot environment for building the packages, it installs the last mainline ArchLinux packages. So Qubes-builder builds the `qubes-vm-gui` with xorg-server 1.21 (so `'X-ABI-VIDEODRV_VERSION=25.2' 'X-ABI-XINPUT_VERSION=24.4'`).

Remember that you can clone your current arch template, and do experiments without do damages to your reference arch template.

**Comment by seberm** [comment]
Hello @lubellier,

> Remember that you can clone your current arch template, and do experiments without do damages to your reference arch template.

Sure, that's what I do :), but thanks.

I've finnaly got time to rebuild the qubes-vm-gui package in the chroot env using [these steps](https://wiki.archlinux.org/title/DeveloperWiki:Building_in_a_clean_chroot#Classic_way) (thanks for the tip!). I can confirm, that the @jose19's PR (https://github.com/QubesOS/qubes-gui-agent-linux/pull/146) works well on Qubes v4.0.

Thanks to the chroot env, I have also found out that the `lsb-release` make dependecy was missing from the PKGBUILD:
- https://aur.archlinux.org/cgit/aur.git/commit/?h=qubes-gui-agent-linux&id=59e9ca14c830f1b33fd74ed77106860c4e0f75e4

We should probably specify this dependency also in the [PKGBUILD maintained by Qubes](https://github.com/QubesOS/qubes-gui-agent-linux/blob/master/archlinux/PKGBUILD), should I make a PR?

Thanks!

**Comment by lubellier** [comment]
Hello @sebterm ,

> I've finnaly got time to rebuild the qubes-vm-gui package in the chroot env using these steps (thanks for the tip!). I can confirm, that the @jose19's PR (QubesOS/qubes-gui-agent-linux#146) works well on Qubes v4.0.

Good news :-)

> We should probably specify this dependency also in the PKGBUILD maintained by Qubes, should I make a PR?

Sure, please, do the PR.

**Comment by lubellier** [comment]
@andrewdavidwong @jobforacowboy now a closable issue, Marek released [qubes-gui-agent-linux-4.1.24](https://github.com/QubesOS/qubes-gui-agent-linux/commit/e7fa68432e17387437eff4c078d51ef3d08c1d08) with the [#0fa4e24](https://github.com/QubesOS/qubes-gui-agent-linux/commit/0fa4e24e2d4945156093ddfee35eb60119828509) which solves this issue.

Note to ArchLinux users: qubesbot restarts to publish the archlinux packages, see the [update status with archlinux tag](https://github.com/QubesOS/updates-status/labels/r4.1-archlinux-cur-test) and the [updated repository](https://archlinux.qubes-os.org/r4.1/current-testing/vm/archlinux/pkgs/) with 2022 versions of qubes-db-vm, qubes-vm-core, qubes-vm-gui, ...

**Comment by andrewdavidwong** [comment]
Closing as resolved. If anyone believes this issue is not yet resolved, or if anyone is still affected by this issue, please leave a comment, and we'll be happy to reopen it. Thank you.
